### DISCUSSION TITLE
RFC: Linter

### DISCUSSION THREAD

**Discussion by Boshen** [oxc-project/oxc#38] [RFC]
In this discussion I will document how the linter is going to be implemented.

Prioritizations are performance, simplicity and contribution friendly.
This is an initial draft for the first MVP, which may lack some features, such as module resolution.

I'll use structured comments below for better discussion.

**Comment by Boshen** [comment]
### File Traversal

I will use the [ignore](https://crates.io/crates/ignore) crate for file traversal. `.eslintignore` files can be added by calling [add_custom_ignore_filename](https://docs.rs/ignore/0.4.20/ignore/struct.WalkBuilder.html#method.add_custom_ignore_filename).

**Reply by charliermarsh** [reply]
We use the `ignore` crate. It works well. But we _only_ use it for `.gitignore` support, and not for enforcing Ruff's custom inclusion and exclusions, the primary reason being that we support ESLint-like hierarchical configuration, so the set of exclusions can actually change as you traverse the hierarchy and parse configuration files.

**Reply by Brooooooklyn** [reply]
https://github.com/napi-rs/node-rs/blob/main/packages/deno-lint/src/lib.rs#L137

**Reply by charliermarsh** [reply]
Oh interesting, and it resolves those hierarchically? Ruff can't use that AFAICT since the exclusions are defined in `pyproject.toml` files (and `pyproject.toml` files can extend other `pyproject.toml` files) rather than in a flat `gitignore`-like file (a convention picked up from other tools in the ecosystem), but it definitely looks preferable to having to resolve those manually if it works.

**Reply by strager** [reply]
Does *ignore* support placing file watches? File watches are useful for a watch/server mode. File watching needs to be integrated with tree traversal to be done correctly; if you find all the files then add watches, you open yourself up to race conditions.

**Reply by Boshen** [reply]
`ignore` doesn't do file watches. For this scenario I will integrate oxc into lsp extensions in text editors, which means I won't be using walkdir / ignore for file traversal, and the gitignore / eslintignore logic will be handled and cached differently.

**Comment by Boshen** [comment]
### Traversal

The goal here is to maximize multi-core parallelization, where we want to traverse each file, each lint rule and each AST node by using `par_iter` from the rayon crate.

The algorithm:

* for each file (`par_iter`)
  * parse into AST
  * for this AST
    * do a visiting pass, for this pass
      * build a parent pointing tree using the [indextree](https://crates.io/crates/indextree) crate, the nodes of this tree is untyped [0]. 
      * for semantic analysis, build its scope tree, symbols table and control flow graph
  * for each lint rule (`par_iter`)
    * for each node of `indextree` (`par_iter`)
      * call `LintRule.run`
* collect all diagnostics and emit them
    
[0] The untyped node is this gigantic enum

https://github.com/Boshen/oxc/blob/c8cc809766945e7e791a9d7a305bd0cd774f842d/crates/oxc_ast/src/ast_kind.rs#L5-L12

**Reply by ematipico** [reply]
Sounds like a good plan. Let me ask you this: do you plan to have lint rules that work "multi-files"? For example:

```js
// file a.js
export function foo() {}
```

```js
// file b.js
import { foo_typo } from "a.js" // <- lint rule triggered for `foo_typo`
```

**Reply by Boshen** [reply]
> do you plan to have lint rules that work "multi-files"?

Let's move this to another https://github.com/Boshen/oxc/discussions/38#discussioncomment-5087989

**Reply by charliermarsh** [reply]
We don't do a separate visiting vs. linting pass, but it seems like a nice idea. I'm not a big fan of our current approach to this, it still uses the approach I wrote initially which does a single combined pass that simultaneously builds up the scope tree and calls out to lint rules at the appropriate time. Is the downside of the approach you described above not that you need to store the semantic scope analysis for _every_ node, so that you know what bindings were defined when you go to run the lint rule?

We do use something like `indextree` but it's just manually implemented. The AST outlives the AST checker, so the checker just inserts nodes into an array and stores (e.g.) stacks of parent references as it goes. I don't know if it's the "right" approach but it does mean we do very few allocations in the AST pass (mostly just for bindings and lint violations).

**Reply by Boshen** [reply]
>  Is the downside of the approach you described above not that you need to store the semantic scope analysis for every node, so that you know what bindings were defined when you go to run the lint rule?

>  I don't know if it's the "right" approach but it does mean we do very few allocations in the AST pass (mostly just for bindings and lint violations).

Basically I'm following the eslint / clippy approach, where the linting phase is standalone, and every rule has its logic tucked into a single file, all semantics are built prior and fed into the rule.

I do like your approach though, it's inspiring because I faced the issue of having to allocate huge scope trees and symbol tables, which slowed down things a bit. Maybe I can find a middle ground for this.

**Reply by charliermarsh** [reply]
The downside to Ruff's approach is that the rules and the binding / semantic logic are all tightly coupled. `checkers/ast.rs` is now an enormous file that both handles all the binding tracking _and_ calls out to rules at the appropriate moment in time. It's hard to break up (though we'll have to do it at some point) and hard to parallelize. So, I wouldn't necessarily recommend it, although it has "worked" for us thus far :)

**Reply by jfmengels** [reply]
@charliermarsh I've been thinking about this recently, and another downside with your approach (but tell me if I'm wrong, I'd love to know a solution to this problem!) seems to me that rules can't look ahead in the AST. Something that happens in `elm-review` sometimes (not super often but still) is that for convenience's sake, the rule would like to know like the name of every top-level declared type/function, and the traverse the file with that information in mind. Without this look-ahead, rules would have to hold to some potential data while waiting to know whether something was declared or not, and then at the end report errors or discard the data.

An example of that could be if you want a rule that reports references to non-existing functions (given everything being hoisted, which is at least the case in Elm). If you can do a look-ahead to see which functions are available, then as soon as you find an unknown name, you can report it. If you don't, you need to collect all the references and their location, while collecting the declared names, and at the end report the problems.

So I'm thinking that if you don't do look-aheads, then some rules will be more annoying to write but also potentially be more expensive in memory and/or CPU. That said, it looks like with your approach you could still do look-aheads, but not with the semantic analysis helpers figured out.

**Comment by Boshen** [comment]
### Lint Rule

A Lint rule implements the following trait.

```rust
pub trait Rule: Default {
    const CATEGORY: RuleCategory;

    fn run_once(&self, _ctx: &Semantic) -> Vec<Diagnostic> {
        vec![]
    }

    fn run<'a>(&self, _node: &AstNode<'a>, _ctx: &Semantic<'a>) -> Vec<Diagnostic> {
        vec![]
    }
}
```

These two methods returns a `Vec` because they can return multiple diagnostics.

#### Semantic

The semantic struct wraps everything associated with the AST being called upon. For example scopes, symbols, cfg, trivias etc.

There are lint rules which will only act upon semantic data, so I added a `run_once` method for convenience.

#### AstNode

`AstNode` wraps all the context for this node

```rust
pub type AstNode<'a> = indextree::Node<SemanticNode<'a>>;

pub struct SemanticNode<'a> {
    pub kind: AstKind<'a>,

    pub flags: NodeFlags,

    /// Associated scope
    pub scope_id: ScopeId,

    /// Associated control flow node
    pub flow_id: Option<FlowNodeId>,
}

bitflags! {
    #[derive(Default)]
    pub struct NodeFlags: u8 {
        const Class             = 1 << 0; // If Node is inside a class
        const HasImplicitReturn = 1 << 1; // If function implicitly returns on one of codepaths (initialized by binding)
        const HasExplicitReturn = 1 << 2; // If `AstKind::FunctionBody` has explicit reachable return on one of codepaths
        const HasBreak          = 1 << 3; // if `AstKind::WhileStatement` or `AstKind::DoWhileStatement` or `AstKind::ForStatement` has reachable break statement in body.
        const HasYield          = 1 << 4; // if Function has yield statement in body.
        const HasAwait          = 1 << 5; // if Function has await statement in body.
        const Export            = 1 << 6; // Export Declaration
    }
}
```

**Reply by ematipico** [reply]
It's a bit unclear to me when and how a rule could return multiple diagnostics. I mean, I know it myself, but maybe it's  worth explaining it.

**Reply by strager** [reply]
I'm not certain I am right, but I think you should group things as *analyses* rather than *lint rules*. An analysis can have multiple lint rules (i.e. emit different diagnostics). For example, you might have an analysis pass which does all of the following:

* disallow `new String("x")` (ESLint no-new-wrappers)
* disallow `new Boolean(true)` (ESLint no-new-wrappers)
* disallow `new Number(42)` (ESLint no-new-wrappers)
* disallow `new Symbol("x")` (ESLint no-new-symbol)
* disallow `new Function("x")` (ESLint no-new-func)
* disallow `new Object()` (ESLint no-new-object)

If you make each of these a separate lint rule, that's a lot of overhead (in theory).

If you plan on making lint rules for things like ESLint no-undef, ESLint no-redeclare, and ESLint no-const-assign, then you probably want to share work between these rules. That's hard if each rule is isolated.

If what you're talking a 'lint rule' is what I call an analysis, then I think we're on the same page.

**Reply by strager** [reply]
Having `run` execute for *every* lint rule for *every* AST node sounds **expensive**.

1. Algorithmically, that's O(number-of-lint-rules * number-of-AST-nodes). If I add a new lint rule, I just added a linear scan through the AST.
2. Every call to `run` is polymorphic/dynamic. This means `run` cannot be inlined, and you need to structure your outer loop carefully to avoid CPU branch mispredictions due to the dynamic call. (`for each node { for each rule { run } }` is faster than `for each rule { for each node { run } }`.)
3. Every `run` call will dispatch on `AstNode::kind`, which puts a lot of work on the CPU branch predictor. It's also wasted effort; every `run` call does basically the same thing.
4. If a lint rule implements `run_once` and doesn't need `run`, calling `run` is pointless, but it'll be called for every node anyway.

 I would strongly reconsider the `run` design choice.

**Reply by strager** [reply]
Minor point: Returning `Vec<Diagnostic>` is both less convenient and (slightly) slower than accepting a `&mut DiagnosticCollection` or something. Compare:
```rust
fn run<'a>(&self, node: &AstNode<'a>, _ctx: &Semantic<'a>) -> Vec<Diagnostic> {
    let mut diags: Vec<Diagnostic> = vec![];
    match node.kind {
        ... => {
            if ... {
                diags.push(Diagnostic::new(...));
            }
        }
        ... => {
            if ... {
                diags.push(Diagnostic::new(...));
            }
        }
    }
    diags
}
```

```rust
fn run<'a>(&self, node: &AstNode<'a>, ctx: &Semantic<'a>) {
    match node.kind {
        ... => {
            if ... {
                ctx.report_diagnostic(...);
            }
        }
        ... => {
            if ... {
                ctx.report_diagnostic(...);
            }
        }
    }
}
```

**Reply by Boshen** [reply]
@strager Thank you for the inputs. My procedure gave me a 90 times speedup compared to eslint and I was pretty happy about it the first time I figured this out. It was enough to pitch the project internally.

Now you have mentioned it, I do admit it's expensive and I should push this further and adjust the algorithm to make it run even faster.

**Reply by Boshen** [reply]
> returning Vec<Diagnostic> is both less convenient and (slightly) slower than accepting a &mut DiagnosticCollection or something

Good catch! I think there will be two minor problems, given we are running in parallel:

* I'll need a mutex for the diagnostics array, which will kill performance
* the diagnostics will be out of order for a given file, which is slightly annoying to deal with

**Reply by strager** [reply]
> I'll need a mutex for the diagnostics array, which will kill performance

Why? Each thread or file can maintain its own diagnostics collection.

**Reply by jfmengels** [reply]
If a rule wants to collect information when going from one node to another, how would they do that? Store the data in some global variable?

Example: I want to report calls to `Module.function()`, but only if `Module` corresponds to the import of a specific module (like `import * as Module from "./module")

**Reply by strager** [reply]
> If a rule wants to collect information when going from one node to another, how would they do that? Store the data in some global variable?

You can store state in fields in the struct implementing the `Rule` trait.

**Reply by jfmengels** [reply]
And those would then be relative to the file to be analyzed? The framework would then have to create a new `Rule` for every file, in order not to reset that data. Right?

**Reply by strager** [reply]
> The framework would then have to create a new `Rule` for every file, in order not to reset that data. Right?

Good question. I don't know what @Boshen had in mind.

**Reply by Boshen** [reply]
I've implemented 140 rules internally and I had the exact question after writing 30 rules or so. I was also frustrated and started implementing mutations just like how ESLint is doing. Then I started thinking about the problem domain - ESLint had to do this because it doesn't have access to anything else, but I'm writing a compiler, I have everything! It then hit me: I should just make the semantic analyzer better, all semantic context should be efficient reads, mutations block performance, and rules will need to redo works sometimes, which is what I've been seeing in ESLint.

So in the end I had a slightly more complicated semantic analyzer, but super clean lint rules.

**Comment by Boshen** [comment]
### Authoring Rules

Writing rules looks almost identical to ESLint, except its Rust :smile:

```rust
use diagnostics::Diagnostic;
use ast::AstKind;

use crate::{AstNode, Rule, RuleCategory, Semantic};

/// https://eslint.org/docs/rules/no-debugger
#[derive(Debug, Default)]
pub struct NoDebugger;

impl Rule for NoDebugger {
    const CATEGORY: RuleCategory = RuleCategory::Correctness;

    fn run<'a>(&self, node: &AstNode<'a>, _: &Semantic<'a>) -> Vec<Diagnostic> {
        let AstKind::DebuggerStatement(stmt) = node.get().kind else { return vec![] };
        vec![Diagnostic::NoDebugger(stmt.node.range())]
    }
}

#[test]
fn test() {
    use crate::rules::LintRule;
    use crate::test_util::test;
    let pass = ["var test = { debugger: 1 }; test.debugger;"];

    let fail = ["if (foo) debugger"];

    test(LintRule::NoDebugger(NoDebugger), &pass, &fail);
}
```  

Tests are written inside the same file for convenience.

**Reply by charliermarsh** [reply]
One thing I'm happy about with Ruff is that we have a common `define_violation!` macro that also picks up the `rustdoc` -- this is inspired by Clippy though the implementation is a little different:

```rust
define_violation!(
    /// ## What it does
    /// Checks for `__init__` methods that are turned into generators by the
    /// inclusion of `yield` or `yield from` expressions.
    ///
    /// ## Why is this bad?
    /// The `__init__` method is the constructor for a given Python class,
    /// responsible for initializing, rather than creating, new objects.
    ///
    /// The `__init__` method has to return `None`. By including a `yield` or
    /// `yield from` expression in an `__init__`, the method will return a
    /// generator object when called at runtime, resulting in a runtime error.
    ///
    /// ## Example
    /// ```python
    /// class InitIsGenerator:
    ///     def __init__(self, i):
    ///         yield i
    /// ```
    ///
    /// ## References
    /// * [CodeQL: `py-init-method-is-generator`](https://codeql.github.com/codeql-query-help/python/py-init-method-is-generator/)
    pub struct YieldInInit;
);
```

**Comment by Boshen** [comment]
### Rule Configuration

Configurations are parsed from json into normal structs. They can accessed from the rule. 

```rust
/// https://eslint.org/docs/latest/rules/no-extra-boolean-cast
#[derive(Debug)]
pub struct NoExtraBooleanCast {
    enforce_for_logical_operands: bool,
}

impl Default for NoExtraBooleanCast {
    fn default() -> Self {
        Self { enforce_for_logical_operands: true }
    }
}

impl Rule for NoExtraBooleanCast {
    fn from_json(value: serde_json::Value) -> Self {
        Self {
            enforce_for_logical_operands: value
                .get(0)
                .and_then(|v| v.get("enforceForLogicalOperands"))
                .and_then(|v| v.as_bool())
                .unwrap_or_default(),
        }
    }
}
```

**Reply by ematipico** [reply]
One thing that I didn't like while writing the configuration for Rome's lint rules, was the fact that rules with options needed to be parsed at a later time, and as a consequence the configuration had to be parsed multiple times:

For example, given this meta-configuration:
```json
{
	"linter": {
		"category1": { "noDebugger": { "options": {}  }, "noConsole": { "options": {} }  },
	}
}
```

It would mean that `"options": {}` blocks need to be saved as a `RawValue` and then parsed later when calling the rule. From my point of view, I think it's better to parse the options before.

**Reply by Boshen** [reply]
I also had trouble dealing with json schemas from eslint's configuration operations, some of them are super dynamic, which can't be parsed statically from serde, so I had to come up with the `from_json` method and manually parse them. But I think the code here is a bit cleaner than eslint because its isolated away, where as in eslint its part of the `create` method:

https://github.com/eslint/eslint/blob/f9f195ef12deb114fb86763010a23ea0cb4c78d1/lib/rules/no-magic-numbers.js#L83-L90

**Reply by Boshen** [reply]
> was the fact that rules with options needed to be parsed at a later time, and as a consequence the configuration had to be parsed multiple times

I didn't know about this. For my approach, the rules are just structs, so they will only be constructed once and then passed to the lint runner.

**Reply by ematipico** [reply]
The issue I have is that if there's an issue in the JSON (syntax error or invalid values), you'd only know that only when you actually attempt to run the rule, which is a bit annoying.

**Comment by Boshen** [comment]
### ESLint Configuration Comments

For `eslint-disable` and `eslint-disable-next-line` configuration options, I will use the parser to parse them into a structured data, possibly a `BTreeMap`, something like

```rust
struct Trivias {
  eslint_configuration_comments: BTreeMap<u32, (Rule, Span)>, // keyed by span.start
}
```

The [`BTreeMap::range`](https://doc.rust-lang.org/stable/std/collections/struct.BTreeMap.html#method.range) method can be used to find whether a rule disable an AST node or not.

**Reply by jfmengels** [reply]
Disable comments are the bane of linters! Refuse to implement them!
https://www.youtube.com/watch?v=XjwJeHRa53A

(I'm half serious. We're not supporting them in `elm-review`, but JS and ESLint are probably ill-equipped to not have them)

**Reply by Boshen** [reply]
> Disable comments are the bane of linters! Refuse to implement them! https://www.youtube.com/watch?v=XjwJeHRa53A
> 
> (I'm half serious. We're not supporting them in `elm-review`, but JS and ESLint are probably ill-equipped to not have them)

I've watched all your talks and all your blogs, I stand with your point. 

ESLint spoiled developers to just throw disable comments everywhere, which is what I've been seeing internally. 

Although I still need to implement it so people can transition to Oxc, I may be able to make something better, for example raise an error if an ignored rules is not catching any errors.

**Reply by jfmengels** [reply]
Glad I was able to convince you ❤️

> I may be able to make something better, for example raise an error if an ignored rules is not catching any errors.

I think that's a great thing to have if disable comments are a thing. Ruff does this as well, so it seems like it's definitely doable.

**Comment by Boshen** [comment]
### Code Fix

I will use the most basic text editing approach, mainly due to its simplicity. It's interesting to note that rust clippy also uses the text editing approach.

**Reply by charliermarsh** [reply]
Ruff uses a basic text editing approach. I based it on ESLint's model (rules can return a fix with an applicable range and the content to insert (which could be empty for deletions)). I know ESLint is now talking about supporting AST-aware autofix, but honestly I've been pretty happy with how this has worked out for Ruff. It's very simple and very flexible (I don't fully understand how AST-aware autofix would work with stylistic rules that don't touch the AST at all), and while it can introduce syntax errors in theory, it honestly doesn't come up much in practice.

A few things that've come up in Ruff:

1. Because fixes can conflict, and/or fixing one rule can create a new violation (e.g., removing a function call might lead to an import being unused), we run Ruff iteratively until there are no errors left to fix. It's not that efficient but it works well in practice. It does lead to some oddities whereby (e.g.) the number of errors we fix when run with `--fix` might not match the number of fixable errors that we flag in the first place.
2. We don't have any model for rules that require fixes at multiple codepoints (e.g., "Insert an import at the top of the file, then change this function call to this other function call").
3. Like Clippy, we detect if an autofix introduces a syntax error, revert the fix, and tell the user to file an issue. We do the same with autofix loops (fixing an error that leads to a new error that we then fix that restores the existing error), which should also be "impossible" but can come up if we have a poorly implemented fix or unintentionally-conflicting rules.
4. Related to (2)... in Python, an empty "body" is a syntax error, since Python relies on semantic whitespace rather than brackets to define scopes. So if we have an autofix that _deletes_ a statement, we have to be careful not to leave its parent body _empty_ (if deleting the statement _would_ leave the parent empty, we insert replace it with a `pass` statement). This requires tracking deletions as we lint the file, even though some of those deletions _may_ not ultimately be applied, e.g., if they conflict with other fixes and get deferred to the next fix pass. This is just a weird thing where there are some theoretical bugs and the semantic model is not very good. I don't think you'll have this problem but it bothers me a lot.

**Reply by Boshen** [reply]
🤣 code fixes seems to be harder than I thought, I think I'll remove it from the first MVP release.

I didn't know there are so many corner cases for code fixes, thank you for your valuable input!

**Reply by charliermarsh** [reply]
Haha don't be too intimidated! Sometimes it's good not to think about or even know about all the problems / cases in advance. We certainly didn't ship with support for all of these things, they were added over time.

**Reply by jfmengels** [reply]
`elm-review` is doing similar things as what @charliermarsh described above.

For (1), as soon we find a fix, we re-run the linter (as optimally as we can) to find new errors, and so on until there are no more. As @charliermarsh said, it's probably not the most efficient, but it avoids some weird results.

For (2), fixes for `elm-review` are a list of text edits (one edit being text removal, text insertion or text replacement). Fixes are applied in order from nearest-to-the-end to nearest-to-the-start of the end, and if there are overlap of some of the ranges, then the fix is canceled.
This simple system works greatly for us. At some point, I would like to have multi-file fixes (because the tool supports multi-file analysis).
I'd say that text-editing works very well in practice. If someone creates a AST-to-string library, then you can in practice simulate AST replacements.

I'm not doing (3) though. I've wanted it because it makes sense, but in practice we've never seen this happen so no user has asked for it (but we tend to have less contradictory rules that lead to this result).

About catching invalid fixes, `elm-review`'s inner testing framework is extremely thorough in its checks. Basically, if your rule provides a fix in a given scenario, you have to indicate what you expect the fix to change the code into. And if the fix creates a syntax error or there is a different problem with the fix, tests will let you know. (The tests do a LOT of other things I've never seen other linter tests do, see https://twitter.com/jfmengels/status/1609231045118889984 and https://github.com/jfmengels/elm-review/blob/master/documentation/design/test-module.md)

**Comment by Boshen** [comment]
### eslint-plugin-import

eslint-plugin-import, a.k.a multi-file linting.

This won't be part of the first MVP, but I already have a working solution internally. The code is not really nice to work with so I plan to redesign it later.

The working code currently utilizes module resolution (the Rust version of `enhance-resolve` which I also worked on) to build up `ModuleRecord`s (defined in the ecma262 spec).

What I haven't figure out yet is to run the module resolution procedure in a memory efficient way. Because every module needs to wait for its dependencies to resolve, peek memory can go up to a 10GB in large monorepos.

**Reply by strager** [reply]
10 GB? 🤨 How big is each `ModuleRecord`?

**Reply by Boshen** [reply]
> 10 GB? 🤨 How big is each `ModuleRecord`?

I wrote that too quickly, I meant the whole program will peek to 10GB, because there will be thousands of ASTs waiting for their dependencies to resolve :-(

**Reply by jfmengels** [reply]
I'll be doing a write-up for how multi-file analysis works for `elm-review`, because it's in practice quite a complex feature with a lot of implications. I'll link it here when I'm done with it.

**Reply by jfmengels** [reply]
I finally did it: https://jfmengels.net/multi-file-analysis/
I can now close this tab :sweat_smile:
